File: rust_tools/webcrawler_multithread/src/scrapers/people_first_scraper.rs
use time::OffsetDateTime;

use crate::job_traits::job_scraper_trait::job_scraper_trait;

pub struct PeopleFirstScraper;

impl job_scraper_trait for PeopleFirstScraper {
    fn start_url(&self) -> &str {
        "https://jobs.myflorida.com/go/Science%2C-Technology%2C-Engineering-&-Mathematics/2814200/"
    }

    fn should_visit(&self, url: &str) -> bool {
        if !url.contains("/job/") && !url.contains("/go/") {
            return false;
        }

        let keywords = [
            "analyst",
            "programmer",
            "specialist",
            "data",
            "computer",
            "information",
            "tech",
        ];

        // URL must contain at least one of the keywords to be considered relevant
        if !keywords
            .iter()
            .any(|kw| url.to_lowercase().contains(&kw.to_lowercase()))
        {
            return false; // skip this URL
        }

        true
    }

    fn results_filter(&self, url: &str) -> bool {
        !url.contains("#content") && !url.contains("#") && !url.contains("/go/")
    }
    
    fn name(&self) -> String {
        let now = OffsetDateTime::now_utc();
        format!(
            "people_first_{:04}_{:02}_{:02}_{:02}_{:02}_{:02}.csv",
            now.year(),
            now.month() as u8,
            now.day(),
            now.hour(),
            now.minute(),
            now.second()
        )
    }
}

File: python_tools/concurrent_job_scraper/utils/scrapers/people_first_scraper.py
import pdb
from typing import List

from playwright.async_api import Page

from ..templates.job_scraper import JobScraper


class PeopleFirstJobScraper(JobScraper):
    def __init__(self):
        super().__init__(
            "https://jobs.myflorida.com/go/Science%2C-Technology%2C-Engineering-&-Mathematics/2814200/",
            "PeopleFirst_Jobs"
        )
        
    async def get_links(self, page: Page) -> List[str]:
        return await page.eval_on_selector_all(
            "a",
            """elements => elements.filter(e => 
                e.classList.contains('jobTitle-link') ||
                    (e.title && e.title.includes('Page'))
                )
                .map(e => e.href)
            """
        )

    def should_visit(self, url: str) -> bool:
        if '/job/' not in url and '/go/' not in url:
            return False

        keywords =  [
            "analyst",
            "programmer",
            "specialist",
            "data",
            "computer",
            "information",
            "tech",
        ]
        
        is_tech_job = any(keyword.lower() in url.lower() for keyword in keywords)
        
        return is_tech_job
    
    
    def results_filter(self, url: str) -> bool:
        return '#content' not in url and "#" not in url and '/go/' not in url